### PR TITLE
[SYCL][ABI-break] Remove <iostream> headers from "sycl/sycl.hpp" and related files

### PR DESCRIPTION
**Pull request by raaiq1**
Refactored code base to remove <iostream> dependency from "sycl/sycl.hpp". The main motivation behind it is so confrom by LLVM coding standards (https://llvm.org/docs/CodingStandards.html#include-iostream-is-forbidden) which forbids the inclusion of \<iostream\> in library files. This required the following:

- Replacing `cout` or `cerr` or clog statements with `fprintf()` statements
- Removing `>>` , `<<` operators for sycl classes that interact with `<iostream`> and substituting them with `std::string` cast operator if possible, otherwise a `to_string` method was defined (as the case in backend enum)
- Removing any calls to the operators mentioned above (if a string::cast operator couldn't be defined) and replacing them with `printf`
- Refactoring some class methods to take a `FILE*` instead of an \<iostream\> object, such as `void DeviceBinaryImage::dump(std::ostream &Out) const` under ''sycl/include/CL/sycl/detail/pi.hpp" and `virtual void Command::printDot(std::ostream &Stream) const = 0;` and it's various overloads

### CODE REVIEW

File: sycl/plugins/level_zero/usm_allocator.cpp [hunk 1]
```diff
@@ -284,6 +284,7 @@ static class SetLimits {
     if (PoolTrace < 1)
       return;
 
+
     std::cout << "USM Pool Settings (Built-in or Adjusted by Environment "
                  "Variable)"
               << std::endl;
```
Comment (steffenlarsen) [inline sycl/plugins/level_zero/usm_allocator.cpp:290]: Should this be made a `printf`?

File: sycl/source/detail/plugin_printers.hpp [hunk 2]
```diff
@@ -19,110 +19,108 @@ namespace sycl {
...
 
+
+//Note:Casting val to unsigned long may not work for all function specializations
 template <typename T>
 inline typename std::enable_if<!std::is_pointer<T>::value, void>::type
 print(T val) {
-  std::cout << "<unknown> : " << val << std::endl;
+  printf("<unknown> : ");
+  printf("%lu",(unsigned long)val);
+  printf("\n");
```
Comment (steffenlarsen) [inline sycl/source/detail/plugin_printers.hpp:29]: To my knowledge, this will have an effect on the readability of our traces. @smaslov-intel - Thoughts on how to do this without std::cout?

File: sycl/plugins/esimd_emulator/pi_esimd_emulator.cpp [hunk 3]
```diff
@@ -373,27 +374,26 @@ extern "C" {
...
   }                                                                            \
   return PI_ERROR_INVALID_OPERATION;
 
 #define CONTINUE_NO_IMPLEMENTATION                                             \
   if (PrintPiTrace) {                                                          \
-    std::cerr << "Warning : Not Implemented : " << __FUNCTION__                \
-              << " - File : " << __FILE__;                                     \
-    std::cerr << " / Line : " << __LINE__ << std::endl;                        \
+     fprintf(stderr, "%s%s%s","Warning : Not Implemented : ", __FUNCTION__                \
+              ," - File : " ,__FILE__);                                     \
```
[the remark this answers is not in the document]
Comment (raaiq1) [reply sycl/plugins/esimd_emulator/pi_esimd_emulator.cpp:385]: I just used that style for some reason. I'll fix it for this file

File: sycl/plugins/level_zero/usm_allocator.cpp [hunk 4]
```diff
@@ -21,10 +21,10 @@
 #include <unordered_map>
 #include <utility>
 #include <vector>
+#include <iostream>
```
Comment (cperkinsintel) [inline sycl/plugins/level_zero/usm_allocator.cpp:24]: This will ultimately have to be removed as well.

File: sycl/source/detail/config.hpp [hunk 5]
```diff
@@ -21,6 +21,7 @@
 #include <mutex>
 #include <string>
 #include <utility>
+#include <iostream>
```
Comment (cperkinsintel) [inline sycl/source/detail/config.hpp:24]: This, likewise, will need removal, no?

File: sycl/plugins/esimd_emulator/pi_esimd_emulator.cpp [hunk 6]
```diff
@@ -40,6 +40,7 @@
 #include <string>
 #include <thread>
 #include <utility>
+#include <iostream>
```
Comment (cperkinsintel) [inline sycl/plugins/esimd_emulator/pi_esimd_emulator.cpp:43]: and this too

File: sycl/include/CL/sycl/detail/device_filter.hpp [hunk 7]
```diff
@@ -47,38 +48,40 @@ class device_filter_list {
...
+     Out += "cpu";
+   } else if (this->DeviceType == info::device_type::gpu) {
+     Out += "gpu";
+   } else if (this->DeviceType == info::device_type::accelerator) {
+     Out += "accelerator";
+   } else if (this->DeviceType == info::device_type::all) {
+     Out += "*";
+   } else {
+     Out += "unknown";
+   }
```
[the remark this answers is not in the document]
Comment (raaiq1) [reply sycl/include/CL/sycl/detail/device_filter.hpp:70]: Yep. Sure

File: sycl/plugins/esimd_emulator/pi_esimd_emulator.cpp (hunk 6 above)
Comment (raaiq1) [reply sycl/plugins/esimd_emulator/pi_esimd_emulator.cpp]: I can't remove that for some reason. Otherwise the build fails , even though there's no reference to any <iostream> object within the file

**Comment by raaiq1** [comment]
/verify with https://github.com/intel/llvm-test-suite/pull/1084

**Comment by raaiq1** [comment]
/verify with https://github.com/intel/llvm-test-suite/pull/1084

**Comment by raaiq1** [comment]
/verify with https://github.com/intel/llvm-test-suite/pull/1084

**Comment by raaiq1** [comment]
/verify with https://github.com/intel/llvm-test-suite/pull/1084

**Comment by raaiq1** [comment]
/verify with https://github.com/intel/llvm-test-suite/pull/1084

**Comment by cperkinsintel** [comment]
/verify with https://github.com/intel/llvm-test-suite/pull/1084

**Comment by cperkinsintel** [comment]
/verify with https://github.com/intel/llvm-test-suite/pull/1084

**Review by cperkinsintel — APPROVED** [review]
LGTM.

**Comment by cperkinsintel** [comment]
ping to other reviewers. The 'Jenkins/llvm-test-suite' is passing against the matching PR https://github.com/intel/llvm-test-suite/pull/1084 on the test suite.

**Review by smaslov-intel — APPROVED** [review]

File: sycl/include/sycl/detail/common.hpp [hunk 8]
```diff
@@ -116,14 +116,16 @@ static inline std::string codeToString(pi_int32 code) {
                           "Native API returns: "
 
 #ifndef __SYCL_SUPPRESS_PI_ERROR_REPORT
-#include <iostream>
+#include <cstdio>
```
Comment (bader) [inline sycl/include/sycl/detail/common.hpp:119]: @raaiq1, please, add motivation for this change to the PR description. Right now it just describes what PR does, but doesn't explain why do we need this change.

I guess performance has something to do with it. Am I right?
FYI: https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#slio3-prefer-iostreams-for-io

File: sycl/include/sycl/detail/device_filter.hpp [hunk 9]
```diff
@@ -47,37 +45,45 @@ class device_filter_list {
...
-    Out << "gpu";
-  } else if (Filter.DeviceType == info::device_type::accelerator) {
-    Out << "accelerator";
-  } else if (Filter.DeviceType == info::device_type::all) {
-    Out << "*";
-  } else {
-    Out << "unknown";
+inline device_filter::operator std::string() const {
+  std::string Out{};
+  Out += backend_to_string(this->Backend);
```
Comment (bader) [inline sycl/include/sycl/detail/device_filter.hpp:53]:
```suggestion
  std::string Out = backend_to_string(this->Backend);
```

File: sycl/include/sycl/detail/device_filter.hpp [hunk 10]
```diff
@@ -47,37 +45,45 @@ class device_filter_list {
...
+  Out += ":";
+  switch (this->DeviceType) {
```
Comment (bader) [inline sycl/include/sycl/detail/device_filter.hpp:55]: Just out of curiosity: is `this->` really needed here?

File: sycl/plugins/cuda/pi_cuda.cpp [hunk 11]
```diff
@@ -30,10 +30,12 @@ std::string getCudaVersionString() {
   int driver_version = 0;
   cuDriverGetVersion(&driver_version);
   // The version is returned as (1000 major + 10 minor).
-  std::stringstream stream;
-  stream << "CUDA " << driver_version / 1000 << "."
-         << driver_version % 1000 / 10;
-  return stream.str();
+  std::string versionString;
+  versionString = "CUDA ";
```
Comment (bader) [inline sycl/plugins/cuda/pi_cuda.cpp:34]:
```suggestion
  std::string versionString = "CUDA ";
```

File: sycl/plugins/esimd_emulator/pi_esimd_emulator.cpp [hunk 12]
```diff
@@ -114,7 +115,8 @@ static bool PrintPiTrace = false;
 
 static void PiTrace(std::string TraceString) {
   if (PrintPiTrace) {
-    std::cout << TraceString << std::endl;
+    printf("%s\n", TraceString.c_str());
+    fflush(stdout);
```
Comment (bader) [inline sycl/plugins/esimd_emulator/pi_esimd_emulator.cpp:119]: Why do we need to change anything in source files? The PR title says <iostream> is removed from headers files.

NOTE: in addition to source files we can keep plug-in header files untouched, if we update only header files included into `sycl/sycl.hpp`.
The same applies to all headers in `source` directory. They are not included to `sycl/sycl.hpp` as well.

File: sycl/test/helpers.hpp [hunk 13]
```diff
@@ -6,6 +6,7 @@
 //
 //===----------------------------------------------------------------------===//
 
+#include <iostream>
 #include <sycl/sycl.hpp>
```
Comment (bader) [inline sycl/test/helpers.hpp:10]:
```suggestion
#include <sycl/sycl.hpp>

#include <iostream>
```
Let's include `iostream` after sycl.hpp to catch cases when sycl.hpp uses `iostream` functionality w/o including the header.

File: sycl/test/gdb/iostream_test.cpp [hunk 14]
```diff
@@ -0,0 +1,13 @@
+// UNSUPPORTED: true
```
Comment (bader) [inline sycl/test/gdb/iostream_test.cpp:1]: Why is it here?

File: sycl/test/gdb/iostream_test.cpp [hunk 15]
```diff
@@ -0,0 +1,13 @@
...
+// RUN: %clangxx -fsycl -g %s -o %t.out
+// RUN: gdb %t.out -batch -x %sycl_gdb_iostream 2>&1 | FileCheck %s
+// CHECK: ignore next 9998 hits
+//
+// Tests if <sycl/sycl.hpp> headers include any <iostream> headers
```
Comment (bader) [inline sycl/test/gdb/iostream_test.cpp:6]: I think it can be done w/o running a GDB, but just checking the compiler output.
NOTE: it checks host side of the compilation only, so you don't need to compile with `-fsycl` option.

File: sycl/source/detail/config.cpp [hunk 16]
```diff
@@ -45,41 +43,92 @@ static void initValue(const char *Key, const char *Value) {
...
+      break;
+  }
+}
+// Finds index of the newline character in char array
+int newLineIndexR(const char *string, int lowerBound, int upperBound) {
+  int currentIndex = (upperBound + lowerBound) / 2;
+  int c = string[currentIndex];
+  if (c == '\n') {
+    return currentIndex;
+  }
```
Comment (bader) [inline sycl/source/detail/config.cpp:61]: The code above doesn't use braces for one-statement branches. Please, use consistent code style.
```suggestion
  if (c == '\n')
    return currentIndex;
```

File: sycl/include/sycl/detail/common.hpp (hunk 8 above)
Comment (raaiq1) [reply sycl/include/sycl/detail/common.hpp:119]: The main motivation behind it to abide by llvm coding standards, in this case it's avoid including <iostream> in library files: https://llvm.org/docs/CodingStandards.html#include-iostream-is-forbidden

File: sycl/plugins/esimd_emulator/pi_esimd_emulator.cpp (hunk 12 above)
Comment (raaiq1) [reply sycl/plugins/esimd_emulator/pi_esimd_emulator.cpp:119]: I'm sorry the PR title is a little misleading. This does remove \<iostream\> headers form the source files as well. So according to LLVM coding standards \<iostream\> headers are forbidden in 'library files' (https://llvm.org/docs/CodingStandards.html#include-iostream-is-forbidden). I interpreted 'library files' to include both the headers and source files.

**Comment by raaiq1** [comment]
@bader @cperkinsintel There is an alternative solution @aelovikov-intel suggested. I've made a draft PR for it (https://github.com/intel/llvm/pull/6469). I would like to discuss if it's better to merge this PR or the draft PR

File: sycl/include/sycl/detail/common.hpp (hunk 8 above)
Comment (bader) [reply sycl/include/sycl/detail/common.hpp:119]: Please, add such information to the pull request description. It's important context information for doing code review.

**Comment by bader** [comment]
> @bader @cperkinsintel There is an alternative solution @aelovikov-intel suggested. I've made a draft PR for it (#6469). They both accomplish the same goal but this PR may have some performance improvements but I'm not sure. I would like to discuss if it's better to merge this PR or the draft PR

I think #6469 might be good enough as first step, but eventually it would be great to have some soft of internal logger, which can isolate C or C++ APIs we use to output error/warning/notification information.

```c++
#include "dpcpp_logger.h"

ERR(<args>);                  // resolves into std::cerr/clog or fprintf(stderr, ...)
INFO(<args>)/WARN(<args>);    // resolves into std::cout or fprintf(stdout, ...)
```

I think this can be done with zero run-time overhead (i.e. can as fast as using `std::cerr/cout` or `fprintf` directly) and simplify maintenance.

File: sycl/include/sycl/ext/intel/esimd/simd.hpp [hunk 17]
```diff
@@ -125,9 +125,30 @@ class simd : public detail::simd_obj_impl<
     return sycl::ext::oneapi::experimental::simd<Ty, N1>(base_type::data());
   }
 
+  /// @ingroup sycl_esimd_misc
```
Comment (kbobrovs) [inline sycl/include/sycl/ext/intel/esimd/simd.hpp:128]: this is wrong doc group. just drop it

File: sycl/include/sycl/ext/intel/esimd/simd.hpp [hunk 18]
```diff
@@ -125,9 +125,30 @@ class simd : public detail::simd_obj_impl<
...
+  /// Prints a \c simd object to an output stream.
```
Comment (kbobrovs) [inline sycl/include/sycl/ext/intel/esimd/simd.hpp:129]: this is not what the code is doing now, it rather converts the object to std::string

File: sycl/plugins/esimd_emulator/pi_esimd_emulator.cpp [hunk 19]
```diff
@@ -609,8 +611,9 @@ pi_result piDeviceGetInfo(pi_device Device, pi_device_info ParamName,
     /// ESIMDEmuPluginInterfaceVersion : 0.a.b
     /// a : ESIMDEmuPluginInterfaceVersion
     /// b : ESIMDEmuPluginDataVersion
-    sprintf(ESimdEmuVersionString, "0.%d.%d", ESIMDEmuPluginInterfaceVersion,
-            ESIMDEmuPluginDataVersion);
+    sprintf(ESimdEmuVersionString, "0.%lld.%lld",
+            (long long int)ESIMDEmuPluginInterfaceVersion,
```
Comment (kbobrovs) [inline sycl/plugins/esimd_emulator/pi_esimd_emulator.cpp:615]: why this change is needed? the *Version are #define's, so %d should work OK - ?

File: sycl/source/esimd_emulator_device_interface.cpp [hunk 20]
```diff
@@ -62,13 +64,14 @@ __SYCL_EXPORT ESIMDDeviceInterface *getESIMDDeviceInterface() {
...
-              << "library is behind required device interface version"
-              << std::endl
-              << "Found version : " << Interface->version << std::endl
-              << "Required version :" << ESIMD_DEVICE_INTERFACE_VERSION
-              << std::endl;
+    fprintf(stderr,
+            "%s\n"
+            "The device interface version provided from plug-in"
+            "library is behind required device interface version\n"
+            "Found version : %llu\n"
```
Comment (kbobrovs) [inline sycl/source/esimd_emulator_device_interface.cpp:71]: %llu does not seem to be right for printing uintptr_t. See e.g. https://stackoverflow.com/questions/5795978/string-format-for-intptr-t-and-uintptr-t

**Comment by raaiq1** [comment]
This PR is being closed, as it touches a lot of files and prone to more errors. An alternative PR has been opened for the simpler solution (https://github.com/intel/llvm/pull/6469)
